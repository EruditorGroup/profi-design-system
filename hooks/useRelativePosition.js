import { useCallback, useEffect, useLayoutEffect, useMemo, useState } from 'react';
/**
 * Returns style that relative to passed [element] or undefined if element is not provider
 * @param elementRef DOM element react reference
 * @example
 * const parent = useRef<HTMLDivElement>();
 * const style = useRelativePosition(parent);
 * <>
 *  <div ref={parent}>I'm parent</div>
 *  <div style={style}>I have relative position to "parent"</div>
 * </>
 */

export default function useRelativePosition(element, align) {
  const [clientRects, setClientRects] = useState();
  const [css, setCSS] = useState();
  const calcRects = useCallback(() => setClientRects(element === null || element === void 0 ? void 0 : element.getBoundingClientRect()), [element]);
  useLayoutEffect(() => calcRects(), [calcRects]);
  useEffect(() => {
    window.addEventListener('resize', calcRects, false);
    return () => window.removeEventListener('resize', calcRects, false);
  }, [calcRects]);
  useLayoutEffect(() => {
    if (clientRects) {
      const {
        top,
        left,
        right,
        width,
        height
      } = clientRects;
      setCSS({
        width,
        top: top + height
      });
      if (align === 'left') setCSS(css => ({ ...css,
        left
      }));
      if (align === 'right') setCSS(css => ({ ...css,
        right: right - width
      }));
    } else {
      setCSS(undefined);
    }
  }, [clientRects, align]);
  return useMemo(() => {
    if (!css) return undefined;
    return {
      position: 'absolute',
      top: `${css.top}px`,
      left: css.left ? `${css.left}px` : 'auto',
      right: css.right ? `${css.right}px` : 'auto',
      minWidth: `${css.width}px`,
      zIndex: 10000000
    };
  }, [css]);
}