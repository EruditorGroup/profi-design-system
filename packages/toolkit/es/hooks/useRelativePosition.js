import { useCallback, useEffect, useState } from 'react';

/**
 * Returns style that relative to passed [element] or undefined if element is not provider
 * @param elementRef DOM element react reference
 * @example
 * const parent = useRef<HTMLDivElement>();
 * const [style, recalc] = useRelativePosition(parent);
 * <>
 *  <div ref={parent}>I'm parent</div>
 *  <div style={style}>I have relative position to "parent"</div>
 * </>
 */
export default function useRelativePosition(element, align, offset) {
  const [clientRects, setClientRects] = useState();
  const [css, setCSS] = useState();
  const calcRects = useCallback(() => {
    setClientRects(element ? element.getBoundingClientRect() : undefined);
  }, [element]);
  useEffect(() => calcRects(), [calcRects]);
  useEffect(() => {
    window.addEventListener('resize', calcRects, false);
    return () => {
      window.removeEventListener('resize', calcRects, false);
    };
  }, [calcRects]);
  useEffect(() => {
    if (clientRects) {
      const {
        top,
        left,
        width,
        height
      } = clientRects;
      setCSS(css => {
        var _offset$top;

        return { ...css,
          width,
          top: top + height + ((_offset$top = offset === null || offset === void 0 ? void 0 : offset.top) !== null && _offset$top !== void 0 ? _offset$top : 0) + window.scrollY
        };
      });
      if (align === 'left') setCSS(css => {
        var _offset$left;

        return { ...css,
          left: left + ((_offset$left = offset === null || offset === void 0 ? void 0 : offset.left) !== null && _offset$left !== void 0 ? _offset$left : 0) - window.scrollX
        };
      });
      if (align === 'right') setCSS(css => {
        var _offset$right;

        return { ...css,
          left: left - ((_offset$right = offset === null || offset === void 0 ? void 0 : offset.right) !== null && _offset$right !== void 0 ? _offset$right : 0) // right: right - width + (offset?.right ?? 0) + window.scrollX,

        };
      });
    } else {
      setCSS(undefined);
    }
  }, [clientRects, align, offset]);
  if (!css) return [{
    display: 'none'
  }, calcRects];
  return [{
    position: 'absolute',
    top: `${css.top}px`,
    left: css.left ? `${css.left}px` : 'auto',
    right: css.right ? `${css.right}px` : 'auto',
    minWidth: `${css.width}px`,
    zIndex: '10000000' //TODO: maybe control it from props?

  }, calcRects];
}