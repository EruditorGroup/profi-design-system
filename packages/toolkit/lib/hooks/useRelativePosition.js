"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useRelativePosition;

var _react = require("react");

/**
 * Returns style that relative to passed [element] or undefined if element is not provider
 * @param elementRef DOM element react reference
 * @example
 * const parent = useRef<HTMLDivElement>();
 * const [style, recalc] = useRelativePosition(parent);
 * <>
 *  <div ref={parent}>I'm parent</div>
 *  <div style={style}>I have relative position to "parent"</div>
 * </>
 */
function useRelativePosition(element, align, offset) {
  const [clientRects, setClientRects] = (0, _react.useState)();
  const [css, setCSS] = (0, _react.useState)();
  const calcRects = (0, _react.useCallback)(() => {
    setClientRects(element ? element.getBoundingClientRect() : undefined);
  }, [element]);
  (0, _react.useEffect)(() => calcRects(), [calcRects]);
  (0, _react.useEffect)(() => {
    window.addEventListener('resize', calcRects, false);
    return () => {
      window.removeEventListener('resize', calcRects, false);
    };
  }, [calcRects]);
  (0, _react.useEffect)(() => {
    if (clientRects) {
      const {
        top,
        left,
        width,
        height
      } = clientRects;
      setCSS(css => {
        var _offset$top;

        return { ...css,
          width,
          top: top + height + ((_offset$top = offset === null || offset === void 0 ? void 0 : offset.top) !== null && _offset$top !== void 0 ? _offset$top : 0) + window.scrollY
        };
      });
      if (align === 'left') setCSS(css => {
        var _offset$left;

        return { ...css,
          left: left + ((_offset$left = offset === null || offset === void 0 ? void 0 : offset.left) !== null && _offset$left !== void 0 ? _offset$left : 0) - window.scrollX
        };
      });
      if (align === 'right') setCSS(css => {
        var _offset$right;

        return { ...css,
          left: left - ((_offset$right = offset === null || offset === void 0 ? void 0 : offset.right) !== null && _offset$right !== void 0 ? _offset$right : 0) // right: right - width + (offset?.right ?? 0) + window.scrollX,

        };
      });
    } else {
      setCSS(undefined);
    }
  }, [clientRects, align, offset]);
  if (!css) return [{
    display: 'none'
  }, calcRects];
  return [{
    position: 'absolute',
    top: `${css.top}px`,
    left: css.left ? `${css.left}px` : 'auto',
    right: css.right ? `${css.right}px` : 'auto',
    minWidth: `${css.width}px`,
    zIndex: '10000000' //TODO: maybe control it from props?

  }, calcRects];
}