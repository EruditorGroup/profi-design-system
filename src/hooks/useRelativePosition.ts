import {useCallback, useEffect, useState} from 'react';
import type {CSSProperties} from 'react';

export type RelativePositionOffset = {
  top?: number;
  bottom?: number;
  left?: number;
  right?: number;
};

/**
 * Returns style that relative to passed [element] or undefined if element is not provider
 * @param elementRef DOM element react reference
 * @example
 * const parent = useRef<HTMLDivElement>();
 * const style = useRelativePosition(parent);
 * <>
 *  <div ref={parent}>I'm parent</div>
 *  <div style={style}>I have relative position to "parent"</div>
 * </>
 */
export default function useRelativePosition<T extends HTMLElement>(
  element: T | null | undefined,
  align: 'left' | 'right',
  offset?: RelativePositionOffset,
): {[key: string]: string} | undefined {
  const [clientRects, setClientRects] = useState<DOMRect>();
  const [css, setCSS] = useState<CSSProperties>();

  const calcRects = useCallback(
    () => setClientRects(element?.getBoundingClientRect()),
    [element],
  );

  useEffect(() => calcRects(), [calcRects]);

  useEffect(() => {
    window.addEventListener('resize', calcRects, false);
    return () => {
      window.removeEventListener('resize', calcRects, false);
    };
  }, [calcRects]);

  useEffect(() => {
    if (clientRects) {
      const {top, left, right, width, height} = clientRects;
      setCSS((css) => ({
        ...css,
        width,
        top: top + height + (offset?.top ?? 0) + window.scrollY,
      }));
      if (align === 'left')
        setCSS((css) => ({
          ...css,
          left: left + (offset?.left ?? 0) - window.scrollX,
        }));
      if (align === 'right')
        setCSS((css) => ({
          ...css,
          right: right - width + (offset?.right ?? 0) + window.scrollX,
        }));
    } else {
      setCSS(undefined);
    }
  }, [clientRects, align, offset]);

  if (!css) return undefined;
  return {
    position: 'absolute',
    top: `${css.top}px`,
    left: css.left ? `${css.left}px` : 'auto',
    right: css.right ? `${css.right}px` : 'auto',
    minWidth: `${css.width}px`,
    zIndex: '10000000', //TODO: maybe control it from props?
  };
}
