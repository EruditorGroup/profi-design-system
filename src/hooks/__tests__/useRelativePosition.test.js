// yarn test src/client/hooks/__tests__/useRelativePosition.js
import {act, renderHook} from '@testing-library/react-hooks';
import useRelativePosition from '../useRelativePosition';

jest.unmock('../useRelativePosition');

describe('useRelativePosition', () => {
  let element;
  let expectedResult;
  beforeEach(() => {
    element = {
      getBoundingClientRect: jest.fn(() => ({
        top: 20,
        height: 40,
        left: 50,
        width: 70,
      })),
    };
  });

  it('should be defined', () => {
    expect(useRelativePosition).toBeDefined();
  });

  it('should return undefined when element is undefined', () => {
    let hook = renderHook(() => useRelativePosition());
    expect(hook.result.current).toBe(undefined);
  });

  it('should use getBoundingClientRect to calculate styles', () => {
    const hook = renderHook(() => useRelativePosition(element));
    expect(hook.result.current).toMatchSnapshot();
  });
});
