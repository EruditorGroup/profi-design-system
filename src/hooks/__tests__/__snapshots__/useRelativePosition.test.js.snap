// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRelativePosition should use getBoundingClientRect to calculate styles 1`] = `
Object {
  "left": "auto",
  "minWidth": "70px",
  "position": "absolute",
  "right": "auto",
  "top": "60px",
  "zIndex": 10000000,
}
`;
